s3fs/core.py

```diff
@@ -1382,30 +1396,30 @@ def close(self):
                     UploadId=self.mpu['UploadId'],
                     MultipartUpload=part_info)
             else:
-                self.buffer.seek(0)
                 try:
                     self._call_s3(
                         self.s3.s3.put_object,
                         Bucket=self.bucket, Key=self.key,
-                        Body=self.buffer.read(), ACL=self.acl)
+                        Body=self.buffer.getvalue(), ACL=self.acl)
                 except (ClientError, ParamValidationError) as e:
                     raise IOError('Write failed: %s' % self.path, e)
             self.s3.invalidate_cache(self.path)
             self.parts = []
-            self.buffer = None
-        self.closed = True
+            self.buffer.seek(0)
```

Comment:
I initially left it that way but `super(S3File, self).close()` calls `flush()`, which resulted in an `AttributeError` because the buffer was already closed. Moving the super call earlier in the function caused problems because it sets `closed` before calling `flush()`, so we'd lose data.

Reply:
There is not any reason to call `super(S3File, self).close()` at all, or indeed for S3File to derrive from BufferedIOBase.

Reply:
> There is not any reason to call `super(S3File, self).close()` at all

`self.closed` is protected so we can't set it ourselves. It requires a call to the superclass.

> or indeed for S3File to derrive from `BufferedIOBase`

[`TextIOWrapper`](https://docs.python.org/3/library/io.html#io.TextIOWrapper) requires it apparently.

Reply:
You mean this from the [docs](https://docs.python.org/3/library/io.html#io.TextIOWrapper):

> A buffered text stream over a BufferedIOBase binary stream.

This appears to be bad wording, its meaning is that TextIOWrapper derives from BufferedIOBase, not that that object it wraps needs to be one. I have tried, it works if S3File remains a base `object`.

Reply:
Oh, I see.